packages/back-end/src/services/experiments.ts

```diff
@@ -303,11 +303,7 @@ export async function refreshMetric(
 
     let segment: SegmentInterface | undefined = undefined;
     if (metric.segment) {
-      segment =
-        (await SegmentModel.findOne({
-          id: metric.segment,
-          datasource: metric.datasource,
-        })) || undefined;
+      segment = (await findSegmentById(metric.segment, orgId)) || undefined;
```

Comment:
It's fine to not filter by data source in this query directly, but if the data sources don't match, we do want to throw an error.  So maybe just change the check below to

```ts
if (!segment || segment.datasource !== metric.datasource) {
```